sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -318,7 +321,6 @@ trait Unevaluable extends Expression {
  */
 trait RuntimeReplaceable extends UnaryExpression with Unevaluable {
   override def nullable: Boolean = child.nullable
-  override def foldable: Boolean = child.foldable
```

Comment:
This change is not what you'd actually want.
I'm actually really curious why `RuntimeReplaceable` has to implement `Unevaluable` in the first place. It doesn't really make sense. We could just turn `RuntimeReplaceable` into a proper wrapper that delegates everything, right?

Reply:
This change is noop in practice, because we replace `RuntimeReplaceable` at the beginning of the optimizer, so foldable or not doesn't really matter.

Semantically, it might be clearer to just make `RuntimeReplaceable` a pure delegator like `Alias`, but it complicates the expression tree. Anyway, this is out of the scope of this PR.

Reply:
> but it complicates the expression tree

Why would it complicate the expression tree? There's a `TaggingExpression` trait (https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/constraintExpressions.scala#L24) that'd handle this perfectly. Making `RuntimeReplaceable` implement `TaggingExpression` instead of `Unevaluable` makes much more sense to me. It'll still retain the same semantics of being replaced early.

I agree that we can do it in a separate PR, but I'd really like to make sure the refactoring done while we're at it.

Reply:
Ah `TaggingExpression` sounds like a good idea, we can try it later.